src/fcmp_pp/fcmp_pp_types.h

```diff
@@ -67,20 +68,78 @@ struct HeliosT final
+//----------------------------------------------------------------------------------------------------------------------
+// FFI types
+//----------------------------------------------------------------------------------------------------------------------
+// FFI types instantiated on the Rust side must be destroyed back on the Rust side. We wrap them in a unique ptr with a
+// custom deleter that calls the respective Rust destroy fn.
+#define DEFINE_FCMP_FFI_TYPE(raw_t, cpp_fn)                                      \
+    struct raw_t##Deleter { void operator()(raw_t##Unsafe *p) const noexcept; }; \
+    using raw_t = std::unique_ptr<raw_t##Unsafe, raw_t##Deleter>;                \
+    raw_t cpp_fn;
+
+// Sometimes a shared ptr is necessary so that we can reference the same copy of underlying data in multiple places.
+#define DEFINE_FCMP_FFI_SHARED_TYPE(raw_t, cpp_fn1, cpp_fn2) \
+    using raw_t = std::shared_ptr<raw_t##Unsafe>;            \
+    raw_t cpp_fn1;                                           \
+    raw_t cpp_fn2;
+
+// Macro to instantiate an FFI-compatible slice from a vector of FCMP FFI type. Instantiates a vector in local scope
+// so it remains in scope while the slice points to it, making sure memory addresses remain contiguous. The slice is
+// only usable within local scope, hence "TEMP".
+#define MAKE_TEMP_FFI_SLICE(raw_t, vec, slice_name)                              \
+    std::vector<const raw_t##Unsafe *> raw_t##Vector;                            \
+    raw_t##Vector.reserve(vec.size());                                           \
+    for (const raw_t &elem : vec)                                                \
+        raw_t##Vector.push_back(elem.get());                                     \
+    ::raw_t##SliceUnsafe slice_name{raw_t##Vector.data(), raw_t##Vector.size()};
+
+DEFINE_FCMP_FFI_TYPE(HeliosBranchBlind, gen_helios_branch_blind());
+DEFINE_FCMP_FFI_TYPE(SeleneBranchBlind, gen_selene_branch_blind());
+
+DEFINE_FCMP_FFI_TYPE(BlindedOBlind, blind_o_blind(const SeleneScalar &));
+DEFINE_FCMP_FFI_TYPE(BlindedIBlind, blind_i_blind(const SeleneScalar &));
+DEFINE_FCMP_FFI_TYPE(BlindedIBlindBlind, blind_i_blind_blind(const SeleneScalar &));
+DEFINE_FCMP_FFI_TYPE(BlindedCBlind, blind_c_blind(const SeleneScalar &));
+
+DEFINE_FCMP_FFI_TYPE(OutputBlinds,
+    output_blinds_new(const BlindedOBlind &, const BlindedIBlind &, const BlindedIBlindBlind &, const BlindedCBlind &));
+
+DEFINE_FCMP_FFI_SHARED_TYPE(TreeRoot, helios_tree_root(const HeliosPoint &), selene_tree_root(const SelenePoint &));
+
+DEFINE_FCMP_FFI_TYPE(Path,
+    path_new(const OutputChunk &, std::size_t, const HeliosT::ScalarChunks &, const SeleneT::ScalarChunks &));
+
+DEFINE_FCMP_FFI_TYPE(FcmpPpProveInput,
+    fcmp_pp_prove_input_new(const Path &,
+        const OutputBlinds &,
+        const std::vector<SeleneBranchBlind> &,
+        const std::vector<HeliosBranchBlind> &));
```

Comment:
Could these take `epee::span<const SeleneBranchBlind>` and `epee::span<const HeliosBranchBlind>` instead too? They could just as easily to converted to `SeleneBranchBlindSliceUnsafe`, but would avoid extra copies/movies in `tx_builder.cpp`.

Reply:
It's just moving unique pointers in `tx_builder.cpp` / I don't see a way to improve this approach with `epee::span`. I'm open to it though, can you show the code you'd change in `tx_builder.cpp`?

Reply:
Maybe I'll do it in a future PR, not super important.